File: Sources/Gamified/Extensions/DateExtension.swift
//
//  DateExtension.swift
//  
//
//  Created by Nathan Fallet on 27/01/2023.
//

import Foundation

public extension Date {
    
    // Constants
    
    static var previous17Weeks: Date {
        Calendar.current.date(byAdding: .day, value: -118, to: Date()) ?? Date()
    }
    
    static var previous7Weeks: Date {
        Calendar.current.date(byAdding: .day, value: -48, to: Date()) ?? Date()
    }
    
    static var nextYear: Date {
        Calendar.current.date(byAdding: .day, value: 365, to: Date()) ?? Date()
    }
    
    // String for date
    
    var asString: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: self)
    }
    
    // Rendered date
    
    var rendered: String {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .none
        return formatter.string(from: self)
    }
    
    // Date enumeration
    
    var nextStartOfWeek: Date? {
        if Calendar.current.component(.weekday, from: self) == 1 {
            return self
        }
        return Calendar.current.date(byAdding: .day, value: 1, to: self)?.nextStartOfWeek
    }
    
    func getDays(until: Date) -> [Date]? {
        var date = Calendar.current.startOfDay(for: self)
        var result = [Date]()
        
        while date < until {
            result.append(date)
            if let nextDate = Calendar.current.date(byAdding: Calendar.Component.day, value: 1, to: date) {
                date = nextDate
            } else {
                return nil
            }
        }
        return result
    }
    
}

File: Gamified/src/main/java/me/nathanfallet/gamified/extensions/DateExtension.kt
package me.nathanfallet.gamified.extensions

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

val previous17Weeks: Date
    get() {
        val calendar = Calendar.getInstance()
        calendar.time = Date()
        calendar.add(Calendar.DATE, -118)
        return calendar.time
    }

val previous7Weeks: Date
    get() {
        val calendar = Calendar.getInstance()
        calendar.time = Date()
        calendar.add(Calendar.DATE, -48)
        return calendar.time
    }

val previous4Weeks: Date
    get() {
        val calendar = Calendar.getInstance()
        calendar.time = Date()
        calendar.add(Calendar.DATE, -27)
        return calendar.time
    }

val nextYear: Date
    get() {
        val calendar = Calendar.getInstance()
        calendar.time = Date()
        calendar.add(Calendar.DATE, 365)
        return calendar.time
    }

// String for date

val Date.asString: String
    get() {
        val formatter = SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH)
        return formatter.format(this)
    }

// Date enumeration

val Date.nextStartOfWeek: Date
    get() {
        val calendar = Calendar.getInstance()
        calendar.time = this
        while (calendar.get(Calendar.DAY_OF_WEEK) != calendar.firstDayOfWeek) {
            calendar.add(Calendar.DATE, 1)
        }
        return calendar.time
    }

fun Date.getDays(until: Date): List<Date> {
    val calendar = Calendar.getInstance()
    calendar.time = this
    calendar.set(Calendar.HOUR, 0)
    calendar.set(Calendar.MINUTE, 0)
    calendar.set(Calendar.SECOND, 0)
    val result = ArrayList<Date>()

    while (calendar.time < until) {
        result.add(calendar.time)
        calendar.add(Calendar.DATE, 1)
    }

    return result
}
